cfn-resources/teams/cmd/resource/team-user/team_user.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2023 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package teamuser
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util"
+	"github.com/mongodb/mongodbatlas-cloudformation-resources/util/logger"
+	atlasv2 "go.mongodb.org/atlas-sdk/v20231001001/admin"
+)
+
+func FilterOnlyValidUsernames(c TeamUsersAPI, usernames []string) ([]atlasv2.CloudAppUser, *http.Response, error) {
+	var validUsers []atlasv2.CloudAppUser
+	for _, elem := range usernames {
+		userToAdd, httpResp, err := mongoDBCloudUsersAPIClient.GetUserByUsername(context.Background(), elem)
+		if err != nil {
+			_, _ = logger.Warnf("Error while getting the user %s: (%+v) \n", elem, err)
+			return nil, httpResp, err
+		}
+		validUsers = append(validUsers, *userToAdd)
+	}
+	return validUsers, nil, nil
+}
+
+func initUserSet(users []atlasv2.CloudAppUser) map[string]interface{} {
+	usersSet := make(map[string]interface{}, len(users))
+	for _, u := range users {
+		usersSet[u.GetId()] = true
+	}
+	return usersSet
+}
+
+func GetUsersToAddAndRemove(currentUsers []atlasv2.CloudAppUser, newUsers []atlasv2.CloudAppUser) (toAdd []string, toDelete []string, err error) {
+	// Create two sets to store the elements in A and B
+	currentUsersSet := initUserSet(currentUsers)
+	newUsersSet := initUserSet(newUsers)
+
+	// Create two arrays to store the elements to be added and deleted
+	toAdd = []string{}
+	toDelete = []string{}
+
+	// Iterate over the elements in B and add them to the toAdd array if they are not in A
+	for elem := range newUsersSet {
+		if _, ok := currentUsersSet[elem]; !ok {
+			toAdd = append(toAdd, elem)
+		}
+	}
+
+	// Iterate over the elements in A and add them to the toDelete array if they are not in B
+	for elem := range currentUsersSet {
+		if _, ok := newUsersSet[elem]; !ok {
+			toDelete = append(toDelete, elem)
+		}
+	}
+
+	// Return the two arrays
+	return toAdd, toDelete, nil
+}
+
+func UpdateTeamUsers(teamUsersAPIService TeamUsersAPI, existingTeamUsers *atlasv2.PaginatedApiAppUser, usernames []string, orgID, teamID string) error {
```

Comment:
similar to my point on defining interfaces with exactly what you need this method should proabbly be split into the 3 operations it does, (1)validate, (2)remove, (3)add, which would let you have interfaces with only one method in them making them ideal and making your testing simpler

Reply:
I agree in part. Effectively the only (minor) improvement it could be done in this method is to move remove logic into its method

```
for ind := range usersToRemove {
		// remove user from team
		_, err := c.RemoveTeamUser(context.Background(), orgID, teamID, util.SafeString(&usersToRemove[ind]))
		if err != nil {
			return err
		}
	}
```

and add logic into its own
```
for ind := range usersToAdd {
		// add user to team
		newUsers = append(newUsers, atlasv2.AddUserToTeam{Id: util.SafeString(&usersToAdd[ind])})
	}
	// save all new users
	if len(newUsers) > 0 {
		_, _, err = c.AddTeamUser(context.Background(), orgID, teamID, &newUsers)
		if err != nil {
			return err
		}
	}
```

I say minor because the logic of these parts is quite easy. the `UpdateTeamUsers` works as a facade and it represents the business logic of `updating team users in the "teams" resource.

One might argue that `teamUsersAPIService` input param should be three different interfaces: one to get users, one to add and one to remove. I am not 100% sure of the final benefit of it. Maybe the only benefit is on the other methods that will receive their own dedicated interface, but I think the way I might have actually just solved that is by having converted those methods as private, which means they can only be called within this package (hence still be tested).

Ultimately only UpdateTeamUsers is the external interface that should be called and visible.

If you have a strong opinion I can separate them